부산대_FE_한재안 2주차 과제

JaeanHan opened the pull request:
안녕하세요 멘토님!

이번 주차 과제는 페이지 구현 및 라우팅과 session 기반 인증을 구현하는 것이었는데, 
페이지 구현이 오래 걸려 마감 당일에 한 번에 제출하게 됐습니다 ㅠㅠ 
다음 주차부터는 step1, step2, step3 고르게 분류해 제출하겠습니다.

전체적으로 쉽지 않은 주차였습니다.
atomic design, react-router, emotion 모두 다 처음 접해보는 것들이라 월요일에 코딩을 바로 시작할 수 없었습니다.
커밋 로그도 깔끔하게 작성하고 싶었는데 뭔가 우당탕탕 진행한 것 같아 아쉽습니다. 다음 주차엔 깔끔하게 작성하도록 신경쓰겠습니다.

**atomic design**
이번 과제에 atomic design을 처음 사용하며 느낌점은 아래와 같습니다.
 _"쪼개는건 얼마든지 쪼갤 수 있는데 쪼갠 컴포넌트 조각을 다른 컴포넌트를 만들 때 재사용하는 것이 어렵다."_
다른 곳에서도 재사용 가능하도록 확장성 있게 쪼개야하는데, 이러면 그냥 코드만 분리하는게 아니라 형태를 유연하게 고치는 것이 시간을 꽤 많이 잡아먹었습니다. _"이정도면 확장성 있겠지?"_ 라는 생각에 커밋 했는데 나중에 다시 사용하려고 보니 더 유연해야 한다는 것을 깨닫고 수정하기를 계속 반복했습니다.
컴포넌트에 스타일이 들어가면 디자인이 같지 않은 이상 컴포넌트 재사용이 쉽지는 않았던 것 같습니다. 스타일링에 emotion이 아닌 class name을 사용한다면 더 유연한 컴포넌트를 만들 수 있겠다는 생각이 들었습니다. 다음엔 class name을 사용하려고 합니다. 컴포넌트를 어떻게 쪼개면 더 아톰 같을지 여쭙고 싶습니다!

**react router**
공식 홈페이지의 튜토리얼 한 번 따라하고 감을 잡았기 때문에 리액트 라우터 사용 및 적용 자체는 그렇게 어렵지 않았습니다. 
아무래도 처음 사용하는 것이다 보니 기억하고 싶은 기능들을 일부러 사용해 코드의 효율성 측면에서 다소 개선이 필요할 순 있지만, 배움의 측면에선 잘 진행한 것 같습니다. (loader, action등)

로그인 성공시 이전 페이지로 이동과 같은 조건은 어렵지 않았습니다.
다만, 페이지 리다이렉션을 자연스럽게 만드는 것이 어려웠습니다. 
예를 들어서, 
로그안 안했을 때, 마이페이지에 접근하면 로그인 페이지로 이동되어야하고,
이동된 로그인 페이지에서 로그인하면,
다시 마이페이지로 보내져야 하는게 자연스럽게 느껴지는데 이렇게 당연하게 생각되는 리다이렉션 구현이 어려웠습니다.
사실 아직 내부적으로 어떻게 동작하는지 정확하게 몰라서 일단 돌아가게 만들기만 한 상태입니다. ㅠㅠ
`MyPageTemplate.tsx`과 `LoginPage.tsx`에서 `navigate()` 사용을 한 번 봐주시면 좋을 것 같습니다.
_"지금과 같은 상태면 이동은 되더라도 리다이렉션이 여러번 일어나 효율이 좋지 못하다."_ 와 같은 부분이 혹시 있는지, 있다면 왜 그런지가 궁금합니다!

감사합니다!

JaeanHan committed:
feat: add react-router-dom & @types

JaeanHan committed:
chore: atomic design 도입

JaeanHan committed:
chore: stories 파일들 bundle에 포함 되지 않게 외부 stories 폴더로 이동

JaeanHan committed:
chore: .idea 폴더 무시하도록 추가

JaeanHan committed:
feat: flex 관련 속성 props로 받도록 변경

JaeanHan committed:
feat: add Header.tsx

- 왼쪽 선물하기 컴포넌트 추가
- 오른쪽 Profile 컴포넌트 추가

JaeanHan committed:
feat: add Header.tsx - 누락된 Profile.tsx 추가

JaeanHan committed:
feat: Container 프롭스 타입 추가 및 적용

- ContainerJustifyContentType
- ContainerAlignItemsType
- ContainerDirectionType

JaeanHan committed:
feat: Paragraph 아톰 추가 및 Profile에 적용

JaeanHan committed:
chore: 구조 통일 시키기 위해 파일 이름만 index로 변경

JaeanHan committed:
feat: ImageParagraphBanner (molecule) FriendBanner (organism) 추가

JaeanHan committed:
feat: Footer (organism) 추가

JaeanHan committed:
feat: a 태그 기본 스타일 없애기

JaeanHan committed:
feat: GridItemsBox (molecule), ThemeCategories (organism) 추가

- ThemeItem (molecule) 추가

JaeanHan committed:
feat: Heading (atom) 추가

JaeanHan committed:
feat: export HeadingLevel

JaeanHan committed:
feat: ParagraphHeadingDownBanner (molecule), KakaoAiRecommendation (organism) 추가

JaeanHan committed:
feat: Container를 쪼개서 FlatFlex (atom)을 만들고 Container는 modules로 이동 및 Container 임포트 수정

JaeanHan committed:
feat: props 기본 값 설정 아래로 이동

JaeanHan committed:
feat: FlexItemsBox (molecule) 추가

JaeanHan committed:
feat: Heading2 (molecule) 추가

JaeanHan committed:
feat: Grid, Flex 아톰 className 프롭스로 받도록 명시

JaeanHan committed:
chore: 프롭스 이름 변경

JaeanHan committed:
feat: FlexItemsBox 프롭스에 클래스 네임 추가

JaeanHan committed:
feat: 타입 StaticItemProp 추가 및 적용

- 내용이 변하지 않는 static content 전용 iterating key

JaeanHan committed:
feat: 타입 StaticItemProp -> IteratingItemProp 이름 변경, GridItemsBox, FlexItemsBox 아이템들에 map 돌릴 때 key 부여 하도록 수정

-

JaeanHan committed:
feat: 선물하기 필터링 버튼 구현

- ActionSpecificationButton (molecule)
- GroupSpecificationButton (molecule)
- GiftRankFilters (organism)
- useGiftRankButtons
- GiftRankContext

JaeanHan committed:
feat: ThemeItem 카테고리 및 아이템 프롭스에 id 추가

JaeanHan committed:
feat: Footer (organism) css 속성 추가
- box-sizing: border-box

JaeanHan committed:
feat: GiftRankItems (organism) 추가
- 선물 사진들

JaeanHan committed:
feat: 타입 RankingGoodsItems 프롭 내보내기

JaeanHan committed:
feat: Margin0Auto (atom) 추가

JaeanHan committed:
feat: KaKaoAiRecommendation 스타일 수정
- width: 100%
- box-sizing: border-box

JaeanHan committed:
feat: ErrorPage (page) 추가

JaeanHan committed:
feat: HeaderFooterOutlet (template) 추가

JaeanHan committed:
feat: GiftRankBox (template) 추가
- 실시간 급상승 선물랭킹 필터 + 아이템

JaeanHan committed:
feat: HeaderFooter (template) 추가

JaeanHan committed:
feat: MainTemplate (template) 추가

JaeanHan committed:
feat: 라우터 및 path 및 loader 사용해 인덱스 페이지 구현

JaeanHan committed:
feat: Header, Footer, Container maxWidth 되돌리기

JaeanHan committed:
feat: Margin0Auto (atom) max-width 설정 유연하게 수정

JaeanHan committed:
feat: Theme 페이지 구현
- ThemeTemplate
- ThemedBanner
- ThemedItemsSection

JaeanHan committed:
feat: Login 페이지 구현 및 Auth 컨텍스트 추가
-Profile 오른쪽 라벨 내용 유동이게 변경

JaeanHan committed:
feat: 헤더 로그인/내 계정 버튼 navigate 경로 수정

JaeanHan committed:
feat: 마이 페이지 구현 및 라우트 설정
- router match 되지 않는 uri Error Page로 연결
- export defaultAuthValue

JaeanHan committed:
feat: AuthContext 초기화 할 때 세션 값 확인하게 변경

JaeanHan committed:
feat: ThemeItems 라벨 예시 페이지처럼 하나씩 명시

JaeanHan committed:
feat: 로그인 페이지 로그인 된 상태면 /로 리다이렉트

JaeanHan committed:
feat: 마이페이지 로그인 안됐을 때 주소창으로 들어오면 로그인 페이지로, 아니면 인덱스 페이지로 리다이렉트

JaeanHan committed:
feat: AuthContext 초기화 할 때 세션 값 확인하게 변경 -누락본 추가

JaeanHan committed:
feat: router 경로 설정 변경

JaeanHan committed:
feat: Themedbanner에서 허용하는 Theme 제한, uri의 Theme 이상하면 /로 리다이렉트

JaeanHan committed:
feat: 로그인시 인덱스 페이지가 아니라 이전 페이지로 돌아가게 변경

JaeanHan committed:
feat: 마이페이지 주소창으로 들어왔을 때, 로그인하면 마이페이지로 리다이렉트 되도록 수정

JaeanHan committed:
docs: update README.md

wacilpong commented:
@JaeanHan 
> 컴포넌트를 어떻게 쪼개면 더 아톰 같을지 여쭙고 싶습니다!

구조를 보니 열심히 컴포넌트를 어떻게 관리할지 고민하신 흔적이 보입니다. 대단해요!! 😀 다만, 제가 처음에도 말씀드렸듯 컴포넌트 재사용을, 즉 쪼개는 것을 너무 깊이 고민할 필요는 없습니다. 오히려 컴포넌트 재사용에 신경쓸수록(매몰될수록) 더 복잡해질 수도 있습니다. 그렇기에 컴포넌트 단위를 가져가는 자신만의 기준이 더욱 더 중요한데요.

저의 경우에는 내가 작성한 코드/컴포넌트가 우발적 중복은 아닌가 고민하고, 우발적 중복이라고 판단되면 그냥 다른 코드조각/컴포넌트로 가져가는 편입니다. 저의 경험상 대부분의 경우 합친걸 뜯어내는 건 어렵지만, 나중에라도 뜯어져있는걸 합치는 것은 좀더 쉽거든요.

이러한 저의 기준은 클린아키텍쳐라는 책의 아래 부분을 인상깊게 읽고 생긴것이어서, 한번 참고차 알려드리고 싶습니다.

```
작성한 코드가 진짜 중복일까? 혹은 우발적 중복일까?

중복에도 종류가 있다. 그중 하나는 진짜 중복이다.
이 경우 한 인스턴스가 변경되면 동일한 변경을 그 인스턴스의 모든 복사본에 반드시 적용해야 한다.
또 다른 중복은 거짓된 또는 우발적인 중복이다.
중복으로 보이는 두 코드 영역이 각자의 경로로 발전한다면,
즉 서로 다른 속도와 다른 이유로 변경된다면 진짜 중복이 아니다.
```
즉, 유스케이스의 화면 구조가 매우 비슷해서 코드를 통합하고 싶은 유혹에 빠질 때가 많지만, 케이스 자체가 다른 유스케이스로 분리되는 상황이라면 시간이 지나면서 두 화면은 다르게 변화한다는 것입니다. 비슷한 뷰, 비슷한 알고리즘, 심지어 비슷한 DB스키마를 가져도 '자동반사적’으로 중복을 제거해버리면 나중에 이 둘을 분리해내느라 힘들 것입니다.

그래서 재안님도 본인이 작성한 코드에 존재하는 중복들이 완전히 똑같은 화면으로 나아갈 것인지를 판단해보면 어떨까요? 그렇게 봤을 때 atom요소들은 완벽한 중복들로만 구성될 것 같아요. 예를 들어, 컨펌창 내의 '확인/취소'버튼, 인풋박스, 체크박스, 탭 기능을 하는 라벨모양 버튼, (커머스 도메인 한정) 다양한 할인 및 금액을 표시하는 가격영역 박스 등등.. variation이 거의 없는 요소들이 atom이 되지 않을까요?
(물론 이건 저의 생각이고 기준이니 참고만 해주시길 바랍니다!)

wacilpong commented on README.md:
```diff
@@ -1,15 +1,43 @@
 
-[🔗 link](https://edu.nextstep.camp/s/hazAC9xa/ls/4wYFPW1K)
+1. Props Drilling 하기 싫을 때,
+2. UI에 light/black 모드 적용할 때, Context의 value가 바뀔 때마다 의존하는 컴포넌트들 모두 리렌더링 되기 때문에 구조 상관없이 리렌더링 시키기 좋습니다. 실제로 hook을 써서 해봤는데 잘 안됐습니다. 
```
hook을 썼을 때 잘 안됐던 부분을 예시로 삼아서 redux를 사용하는 목적을 좀더 자세히 작성해주실 수 있을까요?

JaeanHan replied:
해당 부분은 지금 당장 답하기에 지식이 부족해 공부를 조금 더 진행 후 수정하겠습니다!

JaeanHan replied:
리드미의 step3 답변들 전체적으로 수정했습니다! 
혹시 hook을 썼을 때 잘 안됐던 이유를 
_"useTheme(readme 기재) 호출 시 각각의 독립적인 상태가 만들어지기 때문에 하나의 상태를 공유하는 것이 아니었다."_ 
로 결론 지었습니다. 혹시 다른 이유라면, 말씀해주시면 감사할 것 같습니다!

사실 제가 redux 사용 경험이 없어 이번에 부랴부랴 공부했습니다.. 
redux엔 middleware와 같이 좋은 기능들도 많지만,
Provider를 통해 하나의 store 객체를 공유하고 결국 subscriber(consumer)가 받는게
Context API를 활용한 상태 공유 방법과 어떤 차이가 있는지 잘 모르겠습니다.

Context API는 값을 공유하기 방법이라 상태와는 무관하기는 해도,
Context API 활용한 상태 공유와 redux는 내부가 구조화 되었다라는 점 말고 크게 차이가 있나? 라는 생각이 듭니다.
혹시 redux 사용 이유의 답으로 이런 내부 구조의 차이에 있는 건지 여쭤 보고 싶습니다..! ㅜㅜ

wacilpong commented on src/components/atoms/Margin0Auto/index.tsx:
```diff
@@ -0,0 +1,16 @@
+  return <FloatDiv limitMaxWidth={limitMaxWidth}>{children}</FloatDiv>;
+};
+
+const FloatDiv = styled.div<Pick<MarginCenterProp, 'limitMaxWidth'>>`
```
요거는 마진을 0으로 하는 단순 박스 div인 것 같은데 왜 이름이 FloatDiv인지 궁금합니다. css에 float속성이 있기 때문에 헷갈려서, 정확히 하기위해 여쭈어봅니다.

JaeanHan replied:
해당 이름은 아래와 같은 이유로 결정했습니다!
1. 해당 div가 둥둥 떠있는 것이 css의 float 속성과 성격이 비슷하다.
2. Float 속성은 앞으로도 사용 안 할 것 같다.

지금 생각해보니 float 속성을 사용하는 div로 생각될 여지가 있는 것 같습니다. 더 직관적인 이름으로 수정하겠습니다.

wacilpong commented on src/components/organisms/GiftRankFilters/index.tsx:
```diff
@@ -0,0 +1,53 @@
+    <Wrapper>
+      <Heading2 hContent={hContent} />
+      <GroupSpecificationFilter
+        flexItems={groupSpecificationButtons}
+        FlexItemComponent={GroupSpecificationButton}
```
넘기는 props가 컴포넌트이더라도, 소문자로 통일 + 더 일반적인 이름으로 넘기면 좀더 직관적으로 읽히지 않을까요?

```suggestion
        items={groupSpecificationButtons}
        component={GroupSpecificationButton}
```

JaeanHan replied:
저는 컴포넌트를 줘야 한다는 것을 명시하기 위해 대문자로 시작했습니다.
그런데 제가 너무 복잡하게 생각했던 것 같습니다.그냥 component를 명시하니 진짜 직관적인 것 같습니다! 
그동안 이름을 부여할 때, 이름만 보고도 어떤 역할을 하는 컴포넌트인지를 명시하고 싶어 구체적인 이름을 부여했습니다. 리뷰를 보고 직관성에 가중치를 더 부여해도 좋겠다는 생각을 했습니다. 진짜 보고 와.. 했습니다.

감사합니다!!

wacilpong commented on src/router.tsx:
```diff
@@ -0,0 +1,86 @@
+export const RouterPath = {
+  root: '/',
+  theme: '/theme/:themeKey',
+  myPage: '/my-account',
```
개인적인 취향일 수도 있으나, 필드명과 path명이 동일하면 검색하기에도, 직관적으로 이해하기에도 더 용이하지 않을까 싶습니다!
```suggestion
  myPage: '/my-page',
```

JaeanHan replied:
> 필드명과 path명이 동일하면 검색하기에도, 직관적으로 이해하기에도 더 용이하지 않을까 싶습니다!

저도 멘토님과 같은 생각입니다! `'myPage' : 'my-page'`가 훨씬 직관적인데요.
과제 구현 조건에 "나의 페이지(`my-account`)"라고 적혀있었어서 어떻게 할 지 고민하다가 그냥 그대로 적었습니다.
지금까지 전체적으로 직관성에 대해 생각을 많이 하게 되는 것 같습니다.
그래도 과제 요구사항에 `my-account`가 명시되어 있으니 myAccount로 변경하겠습니다!

wacilpong commented on src/components/templates/MyPageTemplate/index.tsx:
```diff
@@ -0,0 +1,59 @@
+    setAuth(defaultAuthValue);
+  };
+
+  useEffect(() => {
+    if (auth.name === undefined) {
+      /**
+       * 주소창에 localhost:3000/my-account으로 바로 들어오면 login으로 보내고
+       * localhost:3000/my-account로 들어왔을 때 로그인하면 마이페이지로 리다이렉트 되도록,
+       * 마이페이지에서 로그아웃하면 인덱스 페이지로 보내기
+       */
+      if (isFirstJumpedIn.current) {
+        navigate(RouterPath.login, { state: { from: location.pathname } });
+      } else {
+        navigate(RouterPath.root);
+      }
+    }
+  });
```
이 부분의 경우 마이페이지에 진입해서 체크할 수도 있지만, 로그인 정보는 사실 진입 전부터 알 수 있으므로 여기 로직을 삭제하고 아래처럼 private route 자체를 나눌 수도 있습니다. 이를 통해 불필요한 라우팅을 막을 수 있어요. (아래는 제가 돌아가게만 대충 짠 코드이니 참고만 해주세요.)

```
// PrivateRoute.tsx
import React, { useContext } from "react";
import { Navigate, Outlet } from "react-router-dom";

import { AuthContext } from "./components/contexts/AuthContext";

export const PrivateRoute = (): React.ReactElement => {
  const auth = useContext(AuthContext);

  return auth.name ? <Outlet /> : <Navigate to="/" />;
};

// router.tsx
export const router = createBrowserRouter([
  {
    path: RouterPath.root,
    element: <HeaderFooterOutlet />,
    children: [
      {
        index: true,
        loader: mockImageLoader,
        element: <MainTemplate />,
      },
      {
        path: RouterPath.theme,
        loader: mockImageLoader,
        element: <ThemeTemplate />,
      },
      // 이 부분이 추가됨, 권한처리가 필요한 페이지는 여기에 넣음
      {
        element: <PrivateRoute />,
        children: [
          {
            path: RouterPath.myPage,
            element: <MyPageTemplate />,
          },
        ]
      }
    ],
  },
  {
    path: RouterPath.login,
    action: loginAction,
    element: <LoginPage />,
  },
  {
    path: RouterPath.notFound,
    element: <ErrorPage />,
  },
]);


```

JaeanHan replied:
```
export const PrivateRoute = (): React.ReactElement => {
  const auth = useContext(AuthContext);

  return auth.name ? <Outlet /> : <Navigate to="/" />;
};
```
이처럼 로그인 정보는 페이지 진입 전에도 알 수 있음으로 위에서 먼저 처리하는 것이 기존보다 더 좋은 방법인 것 같습니다! 리액트 라우터 감 잡았다 했는데 구조부터 틀려 조금 부끄럽습니다 ㅎㅎ...
그런데 저는 위의 코드를 보고 아래와 같은 생각을 했습니다.
_"아 저렇게 하면 최상단에서 hook을 쓸 수 있구나!"_
`<Outlet/>`을 활용하여 최상단 라우터를 변형해 제가 원하는 결과물을 한 번 만들어 보겠습니다.

감사합니다!!

wacilpong replied:
코드가 동작한다면 틀렸다고 생각하지 않습니다. 이런식으로도 쓸 수 있고, 좀더 효율적인 방법이 있을 수 있고, 제가 생각한 것보다도 더 효율적이고 '좋은' 방식이 있을 것입니다. 그러니 계속 그 좋은 방식이 뭔지를 찾아보고 생각해보면서 이어나가면 좋겠네요~

라우팅 부분을 수정한 것도 확인해보았는데 router.js안에서 로그인정보에 따른 라우팅을 처리하게 되니 이전보다 좀더 효율적이고 가독성이 좋아진 것 같습니다. 굿입니다.

JaeanHan replied:
코드 개선의 의미를 더 좋은 방식을 찾는 방법으로 생각하는 것이 좋은 거 같습니다.
앞으로는 오류 수정이 아닌 더 좋은 방식을 찾는 중이라고 생각하려고 합니다!!

JaeanHan committed:
docs: step 3 질문 답변 수정

JaeanHan committed:
Update README.md

JaeanHan committed:
Update README.md

JaeanHan committed:
Update README.md

JaeanHan committed:
Update README.md

JaeanHan committed:
feat: FloatDiv -> Margin0AutoItem 컴포넌트 이름 변경

JaeanHan committed:
feat: Margin0AutoItem -> Margin0AutoedItem 컴포넌트 이름 변경

JaeanHan committed:
feat: ThemeKey validation 로직 변경 및 ThemeTemplate로 이동

JaeanHan committed:
feat: ThemeTemplate 주어진 themeKey 유효한지 확인 후 유효하지 않다면 인덱스 페이지로 보내는 로직 간소화

- useEffect-navigate 대신 Navigate 사용

JaeanHan committed:
chore: context 폴더 components 아래에 있었는데 src 아래로 이동

JaeanHan committed:
chore: context 임포트 경로 수정

JaeanHan committed:
feat: 라우팅 구현 완료
- 요구사항 만족과 키-경로 직관성을 위해 키 변경 myPage -> myAccount로 수정

JaeanHan committed:
Update src/components/organisms/GiftRankFilters/index.tsx

Co-authored-by: 루미 <cilpong@gmail.com>

JaeanHan committed:
feat: Flex, Grid 컨테이너 공통 부분 추상화

- FlexItemsBox -> FlexContainer로 변경
- GridItemsBox -> GridContainer로 변경
- 공통부분 IContainer로 추상화
- Container 타이핑 강화
- 기존 Container -> CenteringContainer로 이름 변경 (IContainer와 상관x)

JaeanHan commented:
@wacilpong 
안녕하십니까 멘토님! 말씀해주신 진짜 중복과 우발적 중복의 구분에 대해 고민해봤습니다.😊

> 즉 서로 다른 속도와 다른 이유로 변경된다면 진짜 중복이 아니다.

지금 당장은 같더라도, 추후 갈라질 가능성이 있다면 우발적 중복인 것 같습니다!
코드가 같아도 케이스가 다르다면, 추후 서로 갈라질 가능성이 있기에 진짜 중복으로 판단하면 안될 것 같습니다.
이를 당장 화면에 적용한다면, 절대 변하지 않을 디자인 조각을 기준으로 판단할 수 있다고 생각됩니다.

예를 들어 만약 키보드를 화면에 구현해야 한다면, 키 패드의 크기 별 블록(스페이스 바, 보통 철자 키 등), 그림자/LED 라이트 효과(필요시), 철자/그림(svg)가 아톰인 것 같습니다! 😀

지금 당장 화면을 기준으로 진짜 중복과 우발적 중복을 구분하기엔 어려움이 있어 실습 주제를 살짝 바꿔봤습니다!
바로 진짜 중복과 variation을 분리하는 것입니다. 목표는 진짜 중복을 찾는 것 입니다!

`<Container />` 를 variation(케이스)에 따라 진짜 중복과 variation으로 분리했습니다.
진짜 중복은 `<IContainer />`로 모았고, 각 variation은 `<FlexContainer />`와 `GridContainer />`로 구분했습니다.

```
const IContainer = <T extends ItemProps>({
  items,
  component: Component,
  container: Container,
  otherProps = {},
}: IContainerProps<T>): ReactElement => {
  return (
    <Container {...otherProps}>
      {items.map(({ id, ...props }) => (
        <Component key={id} id={id} {...props} />
      ))}
    </Container>
  );
};

const FlexContainer = <T extends ItemProps>({
  items,
  component,
  className = '',
}: FlexContainerProps<T>): ReactElement => {
  return (
    <IContainer items={items} component={component} container={FlatFlex} otherProps={{ className }} />
  );
};

const GridContainer = <T extends ItemProps>({
  items,
  component,
  columns,
  gap = 0,
}: GridContainerProps<T>): ReactElement => {
  return (
    <IContainer
      items={items}
      component={component}
      container={Grid}
      otherProps={{ className, columns, gap }}
    />
  );
};
```

뭔가 그냥 추상화를 한 번 한 것 같지만😅, 
코드 상 `<IContainer />`는 추후에도 절대 변하지 않아 아톰으로 정의할 수 있다고 생각합니다!
이번 주차도 리뷰해 주신다고 고생 많으셨습니다.
감사합니다!😄

JaeanHan committed:
feat: IContainerProps, FlexContainerProps, GridContainerProps 간소화

JaeanHan committed:
docs: update README.md

wacilpong commented on src/components/atoms/IContainer/index.tsx:
```diff
@@ -0,0 +1,18 @@
+
+import type { IContainerProps, ItemProps } from '@/components/atoms/IContainer/types';
+
+export const IContainer = <T extends ItemProps>({
```
이렇게 네이밍해도 괜찮긴 하지만, 이름 앞의 I는 타입스크립트에서 인터페이스 이름 prefix로 자주 쓰입니다.(개인적으로 저는 이렇게 네이밍하지는 않습니다만) 따라서 좀더 대중적인 컨벤션에 따른다면 이는 지양해도 괜찮을 것 같네요!

https://dev.to/mpotapov/naming-interfaces-in-typescript-iuser-vs-user-2cpn

JaeanHan replied:
이번 기회에 타입스크립트 네이밍 컨벤션에 대해 한 번 찾아 봐야 할 것 같습니다. 해당 이름은 바로 수정했습니다.

감사합니다!

wacilpong commented:
2주차 과제도 정말 수고많으셨습니다! 열심히 고민한 흔적이 정말 많이 보입니다. 🥳

wacilpong commented on README.md:
```diff
@@ -1,15 +1,87 @@
+        return (. . .);
+    }
+
+    useTheme을 통해 하나의 상태를 만들어 공유하고 싶었습니다.
+    상태가 drilling을 통해 연결되어 있지 않더라도,
+    setSeelctedTheme을 통해 state를 바꾸면 useTheme에 저장이 될 것이고,
+    최소한 UIComponent가 리렌더링될 때는 useTheme을 다시 호출하니까 마지막 상태가 반환되겠지 했는데
+    UIComponent가 리렌더 돼도 useTheme의 반환 값은 바뀌지 않았습니다.
+    왜 안되지?라는 생각에 진행이 막혔었습니다.
+
+    이는 useTheme을 호출 할 때 마다 각각의 상태가 만들어져 다 따로 놀게 된다는 것을 알게되었고
+    이후 그냥 Context API를 활용했습니다.
+    Context API를 사용하면 "상단 -> 하단 흐름"과 "같은 provider 중첩x"만 지키면 
+    트리 구조 상관 없이 하나의 값을 공유할 수 있습니다.
+    추가적으로 redux를 사용한다면, 하나의 상태를 내부 store에 저장해 여러 곳에서 공유할 수 있습니다.
+    이때, redux는 one-way data flow 구조로 짜여있고, 
+    동작이 store, action, reducer, subscriber를 통해 정형화되어
+    공유된 상태를 변화시키는 것과 해당 영향 예측을 보다 쉽게 할 수 있습니다.
```
질문에 대한 상세한 답변 좋습니다. 👍 경험을 통해 몸소 습득한 지식이네요. 답변해주신 것처럼 redux같은 라이브러리를 통해 정형화된 패턴으로 상태관리를 일관성 있게 할 수 있습니다. 참고로 리덕스 외에도 zustand, mobx와 같은 라이브러리도 있으니 참고정도만 해보세요.

JaeanHan replied:
이번주차는 과제 제출 마감전, step3를 마지막에 빨리 마무리한다는 생각에 핵심적이다고 생각되는 것만 간단히 적었습니다. 다음 주차부턴 질문을 좀 더 상세하게 적도록 하겠습니다..!

JaeanHan committed:
feat: IContainer -> ItemsIterator 이름 변경
- IContainerProps -> ItemsIteratorProps 이름 변경
